### ISSUE TITLE
Find all dependencies blocking a latest release

### ISSUE THREAD

**Issue by ChrisRackauckas** [KristofferC/RegistryCompatTools.jl#10]
One common question to ask when seeing resolver issues is, what packages are blocking latest releases of something? Usually the issue comes down to `X is blocking StaticArrays 1.0` or something like that, so it would be really nice to have `dependency_blocks("DiffEqBase")`

**Comment by timholy** [comment]
Are you basically saying you want a dependency tree rather than flat list?

**Comment by timholy** [comment]
I ask, because we already have what you're asking about:

```julia
help?> RegistryCompatTools
search: RegistryCompatTools

  RegistryCompatTools makes it easier to discover packages that need [compat] updates.

    •  held_back_by("SomePkg") - which packages are holding back SomePkg?

    •  held_back_packages()["SomePkg"] - which packages does SomePkg hold back?

    •  find_julia_packages_github() - discover which packages I can push to
```

But `held_back_by` returns a list.

**Comment by ChrisRackauckas** [comment]
No I want a list, but I want something that recursively on all dependencies runs `held_back_packages()` and returns a pair of (dependency => held back package) so I can go hunt down all of them to help lift the bounds.

**Comment by ChrisRackauckas** [comment]
Basically, the question would be, on DifferentialEquations.jl which non-latest versions should I expect to be seeing and why? I then want to go to all dependencies or dependencies of dependencies and make a PR for Reexport.jl v1.0 or StaticArrays.jl v1.0 (or these days, bump the CompatHelper PR), but I never quite know what I'm looking for. 

This would also be a nice tool for diagnosing resolver issues, because the answer has almost always been "update the compat on one of these dependencies and see what now works. Now we go fix that compat".

**Comment by timholy** [comment]
Haven't you just described `held_back_by`? https://github.com/KristofferC/RegistryCompatTools.jl/blob/a41f4e347c6ce8f3e5fedcb7c6f0f7ca08be4bba/src/RegistryCompatTools.jl#L148-L161

Of course that's only for a single dependency, but you can separately run `held_back_by("Reexport")` and `held_back_by("StaticArrays")`.

**Comment by ChrisRackauckas** [comment]
It's more than a few steps removed from that though. It's, for all direct and indirect dependencies Y of package X, run held_back_by on Y and filter the result to remove anything that's not an indirect or direct dependency of X, and return a list `Y -> dep` meaning "Y's Project.toml is the reason why using package X will never see dependency Z use the latest version". The goal would then to be to try and get this list empty and have a github action constantly check for such issues.

**Comment by timholy** [comment]
Strictly speaking it's not just the Project.toml, although the *fix* can be there. Take for example:
```julia
julia> using RegistryCompatTools

julia> held_back_by("Reexport")
268-element Vector{String}:
 "ACTRModels"
 "ADI"
 "ARCHModels"
 "ASE"
 "AbstractGPs"
 "AbstractSDRs"
 "AcuteML"
 "AdFem"
 "AgentsPlots"
 "Alpaca"
 ⋮
 "UnitfulAstrodynamics"
 "UnitfulCurrency"
 "ValidatedNumerics"
 "ViscousFlow"
 "Wavelets"
 "Winston"
 "Yao"
 "YaoBase"
 "deSolveDiffEq"
```
OK, `Winston` is one of the packages. Here's its [`Project.toml`](https://github.com/JuliaGraphics/Winston.jl/blob/master/Project.toml). While `Reexport` is a direct dependency, there are no `[compat]` bounds. However, check out the [General registry](https://github.com/JuliaRegistries/General/blame/master/W/Winston/Compat.toml): you can see that RetroCap was used to place bounds that are not present in the Project.toml file.

Of course, a new Winston release could fix that.

But you're right that there seem to be more things holding it back than that:
```sh
tim@diva:~/.julia/registries/General$ fls=$(find . -name "Compat.toml")
tim@diva:~/.julia/registries/General$ grep -l Reexport $fls | wc
    359     359   10714
```
359 != 268. I'm not sure of the reason for the difference. Some packages have already upgraded to 1, but I'm not sure that fully accounts for it.

**Comment by DilumAluthge** [comment]
Yeah you can't just look at the Project.toml file, because of the retro-capping.

**Comment by timholy** [comment]
Does this address your question Chris or is there more? Since I'm not quite getting it the best answer might be a PR (this isn't a complicated package). If this is settled, perhaps we could close.

**Comment by ChrisRackauckas** [comment]
>Does this address your question Chris or is there more? 

That doesn't quite address it. I still don't know how to use these tools to address the problem I was looking at. It's basically the inverse of what you have there. I don't care about the other packages Reexport blocks. What I care about are, what are all of the bottlenecks that DifferentialEquations.jl is blocking? I would want an answer like:

DifferentialEquations.jl is blocking:
Reexport because of dependency X
StaticArrays because of dependency Y

So then I go to repos X and Y and go fix it. In what you showed, I have to already know what the problem is in order to identify the problem. I don't know in advance that Reexport is the problem: I want to catch the potential "hey things aren't updating" issue before ever finding a resolver issue.

I think that function is a start though. It needs to be like: for all direct and indirect dependencies, run `held_back_by` on all of them and see if any direct or indirect dependency is in the list, and then identify which packages have a bound holding something back, and return it as a list of pair. I am not quite sure how to grab that list of indirect dependencies.

**Comment by yurivish** [comment]
I think I just ran into this issue too. 

I am trying to understand why a specific package is not on its most recent version. Running `held_back_packages` gives me a list of 51 packages, none of which are packages that I specifically asked to be installed. 

They are presumably indirect dependencies of packages I did directly install, but I don't know how to connect the chain of dependencies all the way back to the packages I see when I look at `]st`.

**Comment by timholy** [comment]
Does it help to
```julia
pkg> add SomePkg@vX.Y.Z
```
and then figure it out from the error message? That's what I usually do. But of course that only helps for packages you happen to have installed, it doesn't fix it for the entire registry. https://julialang.github.io/Pkg.jl/v1/managing-packages/#conflicts

**Comment by ChrisRackauckas** [comment]
No, that's not the full story because I want to anticipate user's issues. Here's a scenario. I get an issue opened saying "updating gives me an error". I dig in using the method you discussed: oh, it's the combination of JuMP and DiffrentialEquations that leads to the error. I dig in more. The issue is that JuMP requires the newest version of package X, while DiffEq requires DiffEqBase which requires RecursiveArrayTools which hasn't merged its compat on package X (or is just missing a tag 🤦). Now I can solve this issue.

But why do I have to wait until users see a broken case to fix it? Why not just have a function that tells me every potential issue case? In fact, I would like to have CI fail if any such cases show up. That way I know the SciML stack is never at fault for the issue, or at least I know all of the cases we have and which dependencies could be the cause.

**Comment by timholy** [comment]
Understood. It sounds like you could run `held_back_packages` and then depth-first search the graph of its `Pkg.dependencies`.

**Comment by yurivish** [comment]
> Does it help to
> ```
> pkg> add SomePkg@vX.Y.Z
> ```
> and then figure it out from the error message? 

There is no error message and the package upgrades (but will downgrade if I run `]up` again). I think this means the dependency holding `Distances` back is `VegaLite`:

```
(@v1.5) pkg> add Distances@v0.10.0
   Updating registry at `~/.julia/registries/General`
   Updating git-repo `https://github.com/JuliaRegistries/General.git`
  Resolving package versions...
Updating `~/.julia/environments/v1.5/Project.toml`
  [b4f34e82] ↑ Distances v0.9.2 ⇒ v0.10.0
  [dd2c4c9e] ↑ NearestNeighborDescent v0.3.1 ⇒ v0.3.4
  [112f6efa] ↓ VegaLite v2.3.0 ⇒ v1.0.0
Updating `~/.julia/environments/v1.5/Manifest.toml`
  [b99e7846] - BinaryProvider v0.5.10
  [864edb3b] ↑ DataStructures v0.17.20 ⇒ v0.18.8
  [b4f34e82] ↑ Distances v0.9.2 ⇒ v0.10.0
  [1c8ee90f] + IterableTables v1.0.0
  [7d188eb4] ↓ JSONSchema v0.2.0 ⇒ v0.1.1
  [dd2c4c9e] ↑ NearestNeighborDescent v0.3.1 ⇒ v0.3.4
  [efcf1570] ↓ Setfield v0.6.0 ⇒ v0.5.4
  [239c3e63] - Vega v2.0.0
  [112f6efa] ↓ VegaLite v2.3.0 ⇒ v1.0.0
```

If I `]rm VegaLite` and run `]up` again, I see

```
  [b4f34e82] ↑ Distances v0.9.2 ⇒ v0.10.0
  [cd3eb016] ↑ HTTP v0.8.19 ⇒ v0.9.2
  [dd2c4c9e] ↑ NearestNeighborDescent v0.3.1 ⇒ v0.3.4
```

So I think that's right. Curiously, if I reinstall `VegaLite`, Distances downgrades again and the 51 results from `held_back_by("Distances")` include a number of packages I am surprised to see as dependencies of `VegaLite`. Maybe I'm misreading the situation but the [list](https://gist.github.com/yurivish/930e3af8c5212891350d718db521c200) includes `ApproximateBayesianComputing` and `SpatialEcology`.

**Comment by ViralBShah** [comment]
This would be nice to have. I often run into some dependent package blocking an upgrade of a package and have to manually figure out which one it is.
